adminSiteServer/authentication.tsx

```diff
@@ -250,3 +251,125 @@ export async function logInWithCredentials(
 
     throw new Error("Invalid password")
 }
+
+interface TailscaleStatus {
+    Self?: {
+        UserID: string
+        TailscaleIPs: string[]
+    }
+    Peer?: {
+        [key: string]: {
+            UserID: string
+            TailscaleIPs: string[]
+            HostName: string
+            Online: boolean
+        }
+    }
+    User?: {
+        [key: string]: {
+            DisplayName?: string
+        }
+    }
+}
+
+export async function tailscaleAuthMiddleware(
+    req: express.Request,
+    res: express.Response,
+    next: express.NextFunction
+) {
+    // If there's a sessionid in cookies, proceed to `authMiddleware` middleware
+    if (req.cookies["sessionid"]) {
+        return next()
+    }
+
+    // Extract client's IP address
+    const clientIp = getClientIp(req)
+
+    // Get Tailscale IP-to-User mapping
+    const ipToUserMap = await getTailscaleIpToUserMap()
+
+    // Get the Tailscale display name / github username associated with the client's IP address
+    const githubUserName = ipToUserMap[clientIp]
+
+    // Next if user is not found, user can still log in as admin
+    if (!githubUserName) {
+        return next()
+    }
+
+    const user = await db
+        .knexInstance()
+        .table("users")
+        .where({ fullName: githubUserName })
+        .first()
+    if (!user) {
+        console.error(
+            `User with name ${githubUserName} not found in MySQL. Please change your Github profile name to match your MySQL user.`
+        )
+        return next()
+    }
+
+    // Authenticate as the user stored in the token
+    const { id: sessionId } = await logInAsUser(user)
+    res.cookie("sessionid", sessionId, {
+        httpOnly: true,
+        sameSite: "lax",
+        secure: ENV === "production",
+    })
+
+    // Save the sessionid in cookies for `authMiddleware` to log us in
+    req.cookies["sessionid"] = sessionId
+
+    return next()
+}
+
+function getClientIp(req: express.Request): string {
+    let ip =
+        (req.headers["x-forwarded-for"] as string) ||
+        req.socket.remoteAddress ||
+        req.ip
+    if (ip && ip.startsWith("::ffff:")) {
+        ip = ip.replace("::ffff:", "")
+    }
+    return ip
+}
+
+async function getTailscaleIpToUserMap(): Promise<Record<string, string>> {
+    return new Promise((resolve) => {
+        exec("tailscale status --json", (error, stdout) => {
+            if (error) {
+                console.error(`Error getting Tailscale status: ${error}`)
+                return resolve({})
```

Comment:
I think it's better if this `reject`s in an error, and the calling site needs to make sure it catches this (probably just by using `.catch(() => ({}))`).